### ISSUE TITLE
MerkleDb may throw IllegalStateException when taking a snapshot after reconnect

### ISSUE THREAD

**Issue by artemananiev** [hashgraph/hedera-services#5636]
### Description

In some cases, MerkleDb throws the following exception during snapshots:
```
2023-03-13 17:50:51.855 128      ERROR EXCEPTION        <<platform-core: signed-state-file-manager 2>> SignedStateFileWriter: Exception when writing the signed state for round 11060 to disk:
java.lang.IllegalStateException: Cannot snapshot to an existing MerkleDb instance
	at com.swirlds.merkledb.MerkleDb.snapshot(MerkleDb.java:479) ~[swirlds-jasperdb-0.36.1.jar:?]
	at com.swirlds.merkledb.MerkleDb.snapshot(MerkleDb.java:447) ~[swirlds-jasperdb-0.36.1.jar:?]
	at com.swirlds.merkledb.MerkleDbDataSourceBuilder.snapshot(MerkleDbDataSourceBuilder.java:155) ~[swirlds-jasperdb-0.36.1.jar:?]
	at com.swirlds.virtualmap.internal.merkle.VirtualRootNode.detach(VirtualRootNode.java:1100) ~[swirlds-virtualmap-0.36.1.jar:?]
	at com.swirlds.virtualmap.internal.pipeline.VirtualPipeline.lambda$detachCopy$2(VirtualPipeline.java:360) ~[swirlds-virtualmap-0.36.1.jar:?]
	at com.swirlds.virtualmap.internal.pipeline.VirtualPipeline.pausePipelineAndExecute(VirtualPipeline.java:562) ~[swirlds-virtualmap-0.36.1.jar:?]
	at com.swirlds.virtualmap.internal.pipeline.VirtualPipeline.detachCopy(VirtualPipeline.java:359) ~[swirlds-virtualmap-0.36.1.jar:?]
	at com.swirlds.virtualmap.internal.merkle.VirtualRootNode.serialize(VirtualRootNode.java:955) ~[swirlds-virtualmap-0.36.1.jar:?]
	at com.swirlds.virtualmap.VirtualMap.serialize(VirtualMap.java:326) ~[swirlds-virtualmap-0.36.1.jar:?]
	at com.swirlds.common.io.streams.MerkleDataOutputStream.writeSerializableNode(MerkleDataOutputStream.java:67) ~[swirlds-common-0.36.1.jar:?]
	at com.swirlds.common.io.streams.MerkleDataOutputStream.writeInternal(MerkleDataOutputStream.java:84) ~[swirlds-common-0.36.1.jar:?]
	at com.swirlds.common.io.streams.MerkleDataOutputStream.lambda$writeMerkleTree$1(MerkleDataOutputStream.java:175) ~[swirlds-common-0.36.1.jar:?]
	at com.swirlds.common.merkle.iterators.MerkleIterator.forEachRemainingWithIO(MerkleIterator.java:410) ~[swirlds-common-0.36.1.jar:?]
	at com.swirlds.common.io.streams.MerkleDataOutputStream.writeMerkleTree(MerkleDataOutputStream.java:169) ~[swirlds-common-0.36.1.jar:?]
	at com.swirlds.platform.state.signed.SignedStateFileWriter.writeStateFileToStream(SignedStateFileWriter.java:98) ~[swirlds-platform-core-0.36.1.jar:?]
	at com.swirlds.platform.state.signed.SignedStateFileWriter.lambda$writeStateFile$0(SignedStateFileWriter.java:113) ~[swirlds-platform-core-0.36.1.jar:?]
	at com.swirlds.common.io.utility.FileUtils.writeAndFlush(FileUtils.java:289) ~[swirlds-common-0.36.1.jar:?]
	at com.swirlds.platform.state.signed.SignedStateFileWriter.writeStateFile(SignedStateFileWriter.java:112) ~[swirlds-platform-core-0.36.1.jar:?]
	at com.swirlds.platform.state.signed.SignedStateFileWriter.writeSignedStateFilesToDirectory(SignedStateFileWriter.java:128) ~[swirlds-platform-core-0.36.1.jar:?]
	at com.swirlds.platform.state.signed.SignedStateFileWriter.lambda$writeSignedStateToDisk$1(SignedStateFileWriter.java:156) ~[swirlds-platform-core-0.36.1.jar:?]
	at com.swirlds.common.io.utility.FileUtils.executeAndRename(FileUtils.java:253) ~[swirlds-common-0.36.1.jar:?]
	at com.swirlds.common.io.utility.FileUtils.executeAndRename(FileUtils.java:225) ~[swirlds-common-0.36.1.jar:?]
	at com.swirlds.platform.state.signed.SignedStateFileWriter.writeSignedStateToDisk(SignedStateFileWriter.java:155) ~[swirlds-platform-core-0.36.1.jar:?]
	at com.swirlds.platform.state.signed.SignedStateFileManager.lambda$saveSignedStateToDisk$0(SignedStateFileManager.java:203) ~[swirlds-platform-core-0.36.1.jar:?]
	at com.swirlds.common.threading.framework.internal.QueueThreadImpl.waitForItem(QueueThreadImpl.java:249) ~[swirlds-common-0.36.1.jar:?]
	at com.swirlds.common.threading.framework.internal.QueueThreadImpl.doWork(QueueThreadImpl.java:229) ~[swirlds-common-0.36.1.jar:?]
	at com.swirlds.common.threading.framework.internal.StoppableThreadImpl.doWork(StoppableThreadImpl.java:611) ~[swirlds-common-0.36.1.jar:?]
	at com.swirlds.common.threading.framework.internal.StoppableThreadImpl.run(StoppableThreadImpl.java:217) ~[swirlds-common-0.36.1.jar:?]
	at java.lang.Thread.run(Thread.java:833) ~[?:?]
```
It looks somehow related to reconnects.

### Steps to reproduce

I can reproduce it by running ServicesMain from hashgraph/hedera-services repo:

1. Run 3 nodes
2. Wait for the nodes to make a few snapshots to disk, the path is hedera-node/data/saved/com.hedera.services.ServicesMain/<node>/123/<round>
3. Stop the nodes
4. Delete all <round> folders for one of the nodes
5. Start the nodes again, one of them will try to reconnect
6. Sometimes the exception is thrown

### Hedera network

other

### Version

0.36.0

### Operating system

None

**Comment by artemananiev** [comment]
Here is what happens in this bug:

1. Assume there are two virtual maps in an application, A and B. They are usually restored from a saved state on startup. The corresponding data sources are in the default MerkleDb instance, e.g. `swirlds-tmp/0-merkledb`
2. When a reconnect is started, two copies of each VM are created:
  2.1 On the teacher side, a copy is created from `VirtualRootNode.detach()`
  2.2 On the learner side, a copy is created from `VirtualRootNode.setupWithOriginalNode()`
3. Current implementation for both ends up with a call to `MerkleDb.snapshot()` with just a single data source
4. The snapshots are made into new temp folders, like `swirlds-tmp/2-merkledb-A` or `swirlds-tmp/4-merkledb-B`
5. Once reconnect is completed, the teacher closes its data source copy
6. The learner also closes a data source, but not the copy, but the original one
7. At this point, one virtual map in learner's merkle tree uses a data source from one MerkleDb instance (0-merkledb), and another virtual map uses a different instance (2-merkledb)
8. When a snapshot is taken, there are two calls to `VirtualRootNode.detach()`, one for each map
9. JasperDB handles it just fine, it snapshots one table at a time
10. MerkleDb is slightly different. When the first virtual root A is detached, it calls `MerkleDb.snapshot()`, which snapshots the whole database, all tables. Then the second virtual root B is detached, and this time `MerkleDb.snapshot()` finds that there is a MerkleDb instance in the target dir. It checks the list of tables in the target DB (it may be A or A and B) and compares to the list of tables to detach (just B). Since the lists are different, the exception is thrown

To fix it, my proposal is to not use `VirtualDataSourceBuilder.snapshot()` call (which results in `MerkleDb.snapshot()`) during reconnects. Instead, introduce a new method to copy an existing data source. This method will reuse the same MerkleDb instance, to avoid having troubles with snapshots in the future. Note that copies and original data sources will share the same name (data source label), so folder names should be changed. Next, there should be a way to mark one of the data sources as "primary" (which will be included to snapshots), and the other as "secondary" (short lived, deleted on close). During reconnects, the teacher will leave the original data source as primary, as the corresponding virtual map will still be in its merkle tree. The learner will do the opposite, it will request the copy to become primary, as this is the copy that is kept in the merkle tree, not the original virtual map.
